File: sortingAlgo/countingSort/countingSort.py
'''COUNTINGSORT
Counting sort is a sorting technique based on keys between a specific range.
It works by counting the number of objects having distinct key values.
Then doing some arithmetic to calculate the position of each object in the
output sequence.''' 


def countSort(arr):
   output = [0 for i in range(256)]                                   
   count = [0 for i in range(256)]                    # Create a count array initialized with 0
   ans = ["" for _ in arr]                            # as strings are immutable

   for i in arr:
      count[ord(i)] += 1                              # position of character in the output array

      
   for i in range(256):
      count[i] += count[i-1]                          # output character array

  
   for i in range(len(arr)):
      output[count[ord(arr[i])]-1] = arr[i]
      count[ord(arr[i])] -= 1                         # array of sorted charcters

   
   for i in range(len(arr)):
      ans[i] = output[i]
   return ans



#TEST
arr = "Algorithms"
ans = countSort(arr)
print ("Sorted character array is "+str("".join(ans)))

File: sortingAlgo/countingSort/countingSort.lua
--- Sort an array using counting sort algorithm.
--- @param t table
--- @return table
local function countingSort(t)
    local n = #t
    local counts = {}
    for i = 1, n do
        counts[t[i]] = (counts[t[i]] or 0) + 1
    end
    local k = 1
    for i = 1, n do
        if counts[i] then
            for j = 1, counts[i] do
                t[k] = i
                k = k + 1
            end
        end
    end
    return t
end

return countingSort